lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/CheckCompoundPattern.java

```diff
@@ -38,36 +38,33 @@
     }
 
     int flagSep = parts[1].indexOf("/");
-    endChars = (flagSep < 0 ? parts[1] : parts[1].substring(0, flagSep)).toCharArray();
+    endChars = flagSep < 0 ? parts[1] : parts[1].substring(0, flagSep);
     endFlags = flagSep < 0 ? new char[0] : strategy.parseFlags(parts[1].substring(flagSep + 1));
 
     flagSep = parts[2].indexOf("/");
-    beginChars = (flagSep < 0 ? parts[2] : parts[2].substring(0, flagSep)).toCharArray();
+    beginChars = flagSep < 0 ? parts[2] : parts[2].substring(0, flagSep);
     beginFlags = flagSep < 0 ? new char[0] : strategy.parseFlags(parts[2].substring(flagSep + 1));
 
-    replacement = parts.length == 3 ? null : parts[3].toCharArray();
+    replacement = parts.length == 3 ? null : parts[3];
   }
 
   @Override
   public String toString() {
-    return new String(endChars)
-        + " "
-        + new String(beginChars)
-        + (replacement == null ? "" : " -> " + new String(replacement));
+    return endChars + " " + beginChars + (replacement == null ? "" : " -> " + replacement);
   }
 
   boolean prohibitsCompounding(
       CharsRef word, int breakPos, CharsRef stemBefore, CharsRef stemAfter) {
     if (isNonAffixedPattern(endChars)) {
-      if (!charsMatch(word, breakPos - stemBefore.length, stemBefore.chars)) {
```

Comment:
There was a bug here: `stemBefore.chars` was checked fully, disregarding `offset/length`. Now we pass in `CharSequence` to avoid that